File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static List<Integer>[] graph;
	static boolean[] visited;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int v = Integer.parseInt(st.nextToken());

		graph = new ArrayList[n + 1];

		for (int i = 1; i <= n; i++) {
			graph[i] = new ArrayList<>();
		}

		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());

			int x = Integer.parseInt(st.nextToken());
			int y = Integer.parseInt(st.nextToken());

			graph[x].add(y);
			graph[y].add(x);
		}

		for (int i = 1; i <= n; i++) {
			Collections.sort(graph[i]);
		}

		visited = new boolean[n + 1];
		dfs(v);
		System.out.println();

		visited = new boolean[n + 1];
		bfs(v);

	}

	private static void dfs(int node) {
		visited[node] = true;
		System.out.print(node + " ");
		for (int i : graph[node]) {
			if (!visited[i]) {
				dfs(i);
			}
		}
	}

	private static void bfs(int node) {
		ArrayDeque<Integer> queue = new ArrayDeque<>();
		queue.offer(node);
		visited[node] = true;
		while (!queue.isEmpty()) {
			int v = queue.poll();
			System.out.print(v + " ");
			for (int i : graph[v]) {
				if (!visited[i]) {
					queue.offer(i);
					visited[i] = true;
				}
			}
		}
	}
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
import sys

sys.setrecursionlimit(10**6)

n, m, v = map(int, sys.stdin.readline().rstrip().split(" "))
graph = [[] for _ in range(n + 1)]


def dfs(node, visited):
    visited[node] = True
    print(str(node), end=" ")
    for i in sorted(graph[node]):
        if not visited[i]:
            dfs(i, visited)


def bfs(node, visited):
    queue = deque()
    queue.append(node)
    visited[node] = True
    while queue:
        v = queue.popleft()
        print(str(v), end=" ")
        for i in sorted(graph[v]):
            if not visited[i]:
                queue.append(i)
                visited[i] = True


for i in range(m):
    x, y = map(int, sys.stdin.readline().rstrip().split(" "))

    graph[x].append(y)
    graph[y].append(x)

dfs_result = []
visited = [False] * (n + 1)
dfs(v, visited)
print(" ".join(dfs_result))

bfs_result = []
visited = [False] * (n + 1)
bfs(v, visited)
print(" ".join(bfs_result))
